sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java

```diff
@@ -230,26 +249,62 @@
         hasItem(encodingId));
   }
 
+  /**
+   * Verifies that for the given {@code Coder<T>} and values of
+   * type {@code T}, both the structural values are equal and the encoded
+   * bytes are equal, in any {@code Coder.Context}.
+   */
   public static <T> void structuralValueConsistentWithEquals(
       Coder<T> coder, T value1, T value2)
       throws Exception {
-
     for (Coder.Context context : ALL_CONTEXTS) {
       CoderProperties.<T>structuralValueConsistentWithEqualsInContext(
           coder, context, value1, value2);
     }
   }
 
+  /**
+   * Verifies that for the given {@code Coder<T>}, {@code Coder.Context}, and
+   * values of type {@code T}, both the structural values are equal and the
+   * encoded bytes are equal.
+   */
   public static <T> void structuralValueConsistentWithEqualsInContext(
       Coder<T> coder, Coder.Context context, T value1, T value2) throws Exception {
-
     assertEquals(
         coder.structuralValue(value1).equals(coder.structuralValue(value2)),
         Arrays.equals(
             encode(coder, context, value1),
             encode(coder, context, value2)));
   }
 
+  /**
+   * Verifies that for the given {@code Coder<T>}, {@code Coder.Context}, and
+   * values of type {@code T}, the structural values are equal after encoding
+   * and decoding the values, in any {@code Coder.Context}.
+   *
+   * <p>Note this is useful to test the correct implementation of a Coder
+   * structural equality with values that don't implement the equals contract.
+   */
+  public static <T> void structuralValueDecodeEncodeEquals(
+          Coder<T> coder, T value1, T value2)
+          throws Exception {
+    for (Coder.Context context : ALL_CONTEXTS) {
+      CoderProperties.<T>structuralValueDecodeEncodeEqualsInContext(
+              coder, context, value1, value2);
+    }
+  }
+
+  /**
+   * Verifies that for the given {@code Coder<T>}, {@code Coder.Context}, and
+   * values of type {@code T}, the structural values are equal after encoding
+   * and decoding the values.
+   */
+  public static <T> void structuralValueDecodeEncodeEqualsInContext(
+          Coder<T> coder, Coder.Context context, T value1, T value2) throws Exception {
```

Comment:
I think what you want is `structuralValue(a) == structuralValue(decode(encode(a))`. This is a proxy for `a == decode(encode(a))` for things that don't have good `equals` implementations.